File: minimum-size-subarray-sum/Solution209.java
// https://leetcode.com/problems/minimum-size-subarray-sum/

class Solution209 {

    // T: O(n), S: O(1)
    public int minSubArrayLen(int target, int[] nums) {
        int minWinSize = Integer.MAX_VALUE;
        int left = 0, sum = 0;
        for (int right = 0; right < nums.length; right++) {
            sum += nums[right];
            while (sum >= target) {
                int currWinSize = right - left + 1;
                minWinSize = Math.min(minWinSize, currWinSize);
                sum -= nums[left];
                left++;
            }
        }
        return minWinSize != Integer.MAX_VALUE ? minWinSize : 0;
    }

    public static void main(String[] args) {
        var sol = new Solution209();
        var nums1 = new int[] { 2, 3, 1, 2, 4, 3 };
        var nums2 = new int[] { 1, 4, 4 };
        var nums3 = new int[] { 1, 1, 1, 1, 1, 1, 1, 1 };
        System.out.println(sol.minSubArrayLen(7, nums1)); // 2
        System.out.println(sol.minSubArrayLen(4, nums2)); // 1
        System.out.println(sol.minSubArrayLen(11, nums3)); // 0
    }

}

File: minimum-size-subarray-sum/solution209.go
// https://leetcode.com/problems/minimum-size-subarray-sum/

package main

import (
	"fmt"
	"math"
)

// T: O(n), S: O(1)
func minSubArrayLen(target int, nums []int) int {
	minWinSize := math.MaxInt32
	sum, left := 0, 0
	for right, n := range nums {
		sum += n
		for sum >= target {
			currWinSize := right - left + 1
			minWinSize = min(minWinSize, currWinSize)
			sum -= nums[left]
			left++
		}
	}
	if minWinSize == math.MaxInt32 {
		return 0
	}
	return minWinSize
}

func main() {
	nums1 := []int{2, 3, 1, 2, 4, 3}
	nums2 := []int{1, 4, 4}
	nums3 := []int{1, 1, 1, 1, 1, 1, 1, 1}
	fmt.Println(minSubArrayLen(7, nums1))  // 2
	fmt.Println(minSubArrayLen(4, nums2))  // 1
	fmt.Println(minSubArrayLen(11, nums3)) // 0
}
